File: panel/models/pyiodide.py
"""
Defines a custom MathJax bokeh model to render text using MathJax.
"""

from bokeh.core.properties import String
from bokeh.models import LayoutDOM

from ..io.resources import bundled_files
from ..util import classproperty


class PyIodide(LayoutDOM):
    """
    A bokeh model that runs pyiodide scripts.
    """

    code = String()

    __javascript__ = ["https://cdn.jsdelivr.net/pyodide/v0.19.0/full/pyodide.js"]

    __js_skip__ = {'loadPyodide': __javascript__}

    __js_require__ = {
        'paths': {
            'mathjax': "//cdn.jsdelivr.net/pyodide/v0.19.0/full/pyodide"
        },
        'shim': {'loadPyiodide': {'exports': "loadPyodide"}}
    }

File: panel/models/pyiodide.ts
import * as p from "@bokehjs/core/properties"
import {HTMLBox, HTMLBoxView} from "@bokehjs/models/layouts/html_box"

export class PyIodideView extends HTMLBoxView {
  model: PyIodide

  override async lazy_initialize(): Promise<void> {
    super.lazy_initialize()
    if (!(window as any).pyiodide) {
      (window as any).pyiodide = await (window as any).loadPyodide({
	indexURL : "https://cdn.jsdelivr.net/pyodide/v0.19.0/full/"
      });
    }
  }

  render(): void {
    if (!(window as any).pyiodide) { return }
    (window as any).pyiodide.runPython(this.model.code)
  }
}

export namespace PyIodide {
  export type Attrs = p.AttrsOf<Props>
  export type Props = HTMLBox.Props & {
    code: p.Property<string>
  }
}

export interface PyIodide extends PyIodide.Attrs {}

export class PyIodide extends HTMLBox {
  properties: PyIodide.Props

  constructor(attrs?: Partial<PyIodide.Attrs>) {
    super(attrs)
  }

  static __module__ = "panel.models.pyiodide"

  static init_PyIodide(): void {
    this.prototype.default_view = PyIodideView
    this.define<PyIodide.Props>(({String}) => ({
      code: [ String, '']
    }))
  }
}
